---
title: "Bioinformatic Resources: project on lung-adenocarcinoma dataset"
author: "Rayan Slatni Anastasia Santo Alex Callegaro"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE,
                      message=FALSE,
                      tidy.opts = list(width.cutoff = 80),
                      tidy = TRUE)
```

```{r, include=FALSE}
library(fgsea)
library(org.Hs.eg.db)
library(biomaRt)
library(clusterProfiler)
library(enrichplot)
library(ggnewscale)
library(DOSE)
library(pathview)
library(tidyverse)
library(dplyr)
library(edgeR)
library(limma)
library(GenomicFeatures)
library(ggplot2)
library(stringr)
library(tidyverse)
library(MotifDb)
library(seqLogo)
library(PWMEnrich)
library(PWMEnrich.Hsapiens.background)
library(igraph)
```

The aim of this project is to perform several analysis on cancer data, specifically on Lung Adenocarcinoma data, retrieved from 100 patients, including case and control groups.

**TASK 1:** load the RData file

```{r, include=FALSE}
#load("C:/Users/santo/OneDrive/Desktop/Bioinformatics/progetto/Lung_adenocarcinoma.RData")
load("Lung_adenocarcinoma.RData")
dim(c_anno_df) #50 case and 50 controls
dim(r_anno_df) #info gene 62940 genes 
```

Among 100 samples we have information on 62940 expressed genes.

**TASK 2**: extract information only about protein coding. In order to do this we use the *hsapiens_gene_ensembl* dataset to retrieve the information needed to build a query.

```{r}
ensembl <- useEnsembl(biomart = "ensembl",dataset = "hsapiens_gene_ensembl", host = "useast.ensembl.org")
```

```{r}
query1 <- getBM(attributes=c("ensembl_gene_id",
                            "entrezgene_id",
                             "external_gene_name",
                             "gene_biotype"),
                filters=c("ensembl_gene_id"), 
                values=r_anno_df$ensembl_gene_id,
                mart = ensembl)

#extract only protein coding genes
query1_protein_coding <- query1[which(query1$gene_biotype=="protein_coding"),]
```

Filter counts and annotation only for protein coding and remove the presence of possible duplicates.

```{r}
r_anno_df_pc<- r_anno_df[query1_protein_coding$ensembl_gene_id, ]
raw_counts_df_pc<- raw_counts_df[query1_protein_coding$ensembl_gene_id, ]

r_anno_df_pc <- r_anno_df_pc[-which(duplicated(r_anno_df_pc$ensembl_gene_id)),]
raw_counts_df_pc <- raw_counts_df_pc[-which(duplicated(rownames(raw_counts_df_pc))),]
dim(raw_counts_df_pc)
```

We ended up with 22155 genes coding for protein.

**TASK 3**: Perform differential expression analysis using edgeR package and select up and down regulated genes. At first we filter the data in order to keep only the genes with a raw count bigger than 20 in at least 5 cases and 5 controls.

```{r}
count_thr <- 20 #threshold for raw count
repl_thr <- 5 #number of replicates

filter_vec <- apply(raw_counts_df_pc,1,
                    function(y) max(by(y, c_anno_df$condition, function(x)                     sum(x>=count_thr))))
# apply the filter on counts
filter_counts_df <- raw_counts_df_pc[filter_vec>=repl_thr,]

# apply the filter on gene annotation
filter_anno_df <- r_anno_df_pc[rownames(filter_counts_df),]
dim(filter_anno_df)
```

After the filtering we obtain 17289 genes.

The selected data have been then normalized using TMM (trimmed mean of M values) and then used to build a cpm (counts per million) table in order to have normalized expression values.

```{r}
edge_c <- DGEList(counts=filter_counts_df,group=c_anno_df$condition,samples=c_anno_df,genes=filter_anno_df) 

edge_n <- calcNormFactors(edge_c,method="TMM")
cpm_table <- as.data.frame(round(cpm(edge_n),2))
head(cpm_table)
```

At this point we can define the experimental design matrix and use it to calculate the dispersion maximizing the negative binomial likelihood and fit a genewise negative binomial generalized linear model with quasi-likelihood tests.

```{r}
design <- model.matrix(~0+group, data=edge_n$samples)
colnames(design) <- levels(edge_n$samples$group)
rownames(design) <- edge_n$samples$sample
edge_d <- estimateDisp(edge_n,design) #calculate dispersion
edge_f <- glmQLFit(edge_d,design) #fit the model
```

Then we define the contrast according to the two conditions that we want to compare, and define the final model.

```{r}
contro <- makeContrasts("case-control", levels=design) 
edge_t <- glmQLFTest(edge_f,contrast=contro)
```

For the final dataset we introduce some cut-off: -select only genes with p-value 0.01 -select logFC ratio grater than 1.5 for up-regulated genes and lower than -1.5 for down.regulated genes -select logCPM greater than 1.

```{r}
DEGs <- as.data.frame(topTags(edge_t,n=20000,p.value = 0.01,sort.by = "logFC"))

DEGs$class <- "="
#+ - to see if the genes are up or down regulated
DEGs$class[which(DEGs$logCPM>1 & DEGs$logFC>1.5)] = "+"
DEGs$class[which(DEGs$logCPM>1 & DEGs$logFC<(-1.5))] = "-"
DEGs <- DEGs[order(DEGs$logFC,decreasing = T),]

table(DEGs$class)
```

As a result we obtaines 884 down-regulated genes and 1199 up-regulated genes. Now we display a volcano plot of the results

```{r, fig.width = 10, fig.height = 6, fig.align = "center", fig.cap= "Volcano plot: red = up-regulated, green = down-regulated",  echo = FALSE }
input_df <- DEGs
xlabel <- "log2 FC control vs case"
ylabel <- "-log10 p-value"

par(fig=c(0,1,0,1), mar=c(4,4,1,2), mgp=c(2, 0.75, 0))	
plot(input_df$logFC, -log(input_df$PValue,base=10), xlab = xlabel, ylab = ylabel, col = ifelse(input_df$class=="=","grey85", ifelse(input_df$class == "+", "indianred", "olivedrab4")), 
     pch = 20, frame.plot = TRUE, cex = 0.8, main = "Volcano plot")
abline(v=0,lty=2,col="grey20")
```

From the volcano plot we can see that the up-regulated genes have a larger range of value for the log2 FC. We can use a heatmap to visualize the distinction between the two groups clustering with respect to up and down regulated genes. From the dendrogram we can see that the two groups are well clusterized according to the different gene expression.

```{r, fig.width = 10, fig.height = 6,fig.cap= "Heatmap: green = control, yellow = case", fig.align = "center", echo = FALSE}
# creation of vector of colors according to the two groups
cols <- c()
for (i in 1 : 100)
  if (c_anno_df$condition[i] == "control"){
    cols <- append(cols, "chartreuse4")
  } else {
    cols <- append(cols, "burlywood3")
  }

# select color palette
pal <- c("blue","white","red")
pal <- colorRampPalette(pal)(10000)

heatmap(as.matrix(cpm_table[which(rownames(cpm_table)%in%DEGs$ensembl_gene_id[which(DEGs$class!="=")]),]),
        ColSideColors = cols,cexCol = 0.5,margins = c(4,4),
        col=pal,cexRow = 0.2 )
```

**TASK 4**: Perform gene set enrichment analysis using clusterProfiler R package, performing both GO (BP and MF) and WP analysis.

First we add additional information to our genes from ensembl database.

```{r}
convert <- getBM(attributes=c("ensembl_gene_id","entrezgene_id"),
                 filters=c("ensembl_gene_id"), 
                 values=DEGs$ensembl_gene_id,
                 mart = ensembl)

DEGs <- merge(DEGs, convert, by = "ensembl_gene_id")
```

Now we clean the data removing NA and duplicates and split in up and down regulated

```{r}
DEGs <- DEGs[which(!is.na(DEGs$entrezgene_id)),]
DEGs <- DEGs[-which(duplicated(DEGs$entrezgene_id)),]
up_DEGs <- DEGs[DEGs$class == "+",]
down_DEGs <- DEGs[DEGs$class == "-",]
```

Here we perform Enrichment analysis both for up and down regulated genes :

-   Ontology term: biological process

```{r}
up_ego_BP <- enrichGO(gene = up_DEGs$external_gene_name,
                      OrgDb = org.Hs.eg.db,
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pAdjustMethod = "BH",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)

down_ego_BP <- enrichGO(gene = down_DEGs$external_gene_name,
                        OrgDb = org.Hs.eg.db,
                        keyType = 'SYMBOL',
                        ont = "BP",
                        pAdjustMethod = "BH",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05)
                        
```

Visualize the top 10 enriched terms for the up regulated genes :

```{r, fig.width=10, fig.height=6,fig.cap="Dotplot for up regulated genes (BP)", fig.align="center", echo = FALSE}
dotplot(up_ego_BP, showCategory = 10)
```

From the plot above we can see that most of the up-regulated genes are related to pathway connected to the cell cycle, in particular at the nuclear level, like the chromosome segregation and the DNA replication. Thanks to the heatplot we can highlight those genes related to the pathways above. Genes coding for Cell Division Cycle result upregulated for all the pathways.

```{r, fig.width=10, fig.height=6, fig.cap= "Heatplot for upregulated genes (BP)" ,fig.align="center", echo = FALSE}
heatplot(up_ego_BP, showCategory = 6)
```

We can also plot an enrichment map that organizes enriched terms into a network with edges connecting overlapping gene sets. In this way, mutually overlapping gene sets tend to cluster together, making it easy to identify functional module.

```{r, fig.width=10, fig.height=10,fig.cap="Emapplot for up regulated genes (BP)", fig.align="center", echo = FALSE}
x2 <- pairwise_termsim(up_ego_BP) 
emapplot(x2)
```

From this plot we can see a big cluster that include most of the biological process, including: cell cycle checkpoint signalling and chromosome segregation.

Here we visualize the top 10 enriched terms for the down regulated genes

```{r, fig.width=10, fig.height=6,fig.cap="Dotplot for down regulated genes (BP)", fig.align="center", echo = FALSE}
dotplot(down_ego_BP, showCategory = 10)
```

In the plot above we can see that most of the pathways are related to components of the cell structure, like extracellular matrix, cilium movement and microtubule formation.

```{r, fig.width=10, fig.height=6, fig.cap="Heatplot for upregulated genes (BP)", fig.align="center", echo = FALSE}
heatplot(down_ego_BP, showCategory = 6)
```

From the heatmap we can see that the genes belonging to the family ADAMTS (A Disintegrin-Like And Metalloprotease with Thrombospondin Motifs) are involved in the extracelullar organization.

```{r, fig.width=10, fig.height=10, fig.cap="Emapplot for down regulated genes (BP)", fig.align="center", echo = FALSE}
x2 <- pairwise_termsim(down_ego_BP) 
emapplot(x2)
```

From the plot above we can see several small clusters. For example there is one cluster related to inflammatory response, and another one related to angiogenenis and control in vascularization.

-   Ontology term: molecular function

```{r}
up_ego_MF <- enrichGO(gene = up_DEGs$external_gene_name,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "MF",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)

down_ego_MF <- enrichGO(gene = down_DEGs$external_gene_name,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "MF",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
                        
                        
```

Visualize the top 10 enriched terms for the up regulated genes :

```{r, fig.width=10, fig.height=6,fig.cap="Dotplot for up regulated genes (BP)", fig.align="center", echo = FALSE}
dotplot(up_ego_MF, showCategory = 10)
up_ego_MF@result$Description[1:10]
```
From the plot above related to the molecular function for the up-regulated genes, we can see that most of the functions are related to enzimatic activity and structural constituent of chromatin. 

```{r, fig.width=10, fig.height=6, fig.cap= "Heatplot for upregulated genes (BP)" ,fig.align="center", echo = FALSE}
heatplot(up_ego_MF, showCategory = 6)
```
From the heatplot we can notice that many genes coding for histon protein (H2A, H2B, H4) are specific only for the structural constituent of chromatin. 

```{r, fig.width=10, fig.height=10,fig.cap="Emapplot for up regulated genes (BP)", fig.align="center", echo = FALSE}
x2 <- pairwise_termsim(up_ego_MF) 
emapplot(x2)
```
From this plot we can identify several networks for example there are links between the ATP hydrolysis activity, microtuble activity and ATP dependent activity acting on DNA. 

Here we visualize the top 10 enriched terms for the down regulated genes:

```{r, fig.width=10, fig.height=6,fig.cap="Dotplot for down regulated genes (MF)", fig.align="center", echo = FALSE}
dotplot(down_ego_MF, showCategory = 10)
down_ego_MF@result$Description[1:10]
```
In the dotplot for the down regulated genes we can see that most of the functions are related to glycosaminoglycan, sulfur, heparin, cytokin and carbohydrate binding.

```{r, fig.width=10, fig.height=6, fig.cap="Heatplot for upregulated genes", fig.align="center", echo = FALSE}
heatplot(down_ego_MF, showCategory = 6)
```

From the heatmap we can see that the genes belonging to the family ADAMTS, FGFR and RSPO are involved in the heparin, glycosaminoglycan and and sulfur compound binding.

```{r, fig.width=10, fig.height=10, fig.align="center", echo = FALSE}
x2 <- pairwise_termsim(down_ego_MF) 
emapplot(x2)
```
From this plot we can identify several networks for example there are links between the cargo receptor activity and lipoprotein particles binding, and between sulfur compound, glycosaminoglycan and heparin binding
. 
-   KEGG enrichment analysis

```{r}
up_eWP <- enrichWP(gene = up_DEGs$entrezgene_id,
                organism = 'Homo sapiens',
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.1)

down_eWP <- enrichWP(gene = down_DEGs$entrezgene_id,
                organism = 'Homo sapiens',
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.1)

```

Here we visualize the top 10 enriched terms for the down and up regulated genes: 
```{r, fig.width=10, fig.height=6,fig.cap="Dotplot for up regulated genes (BP)", fig.align="center", echo = FALSE}
barplot(down_eWP,showCategory=10)
barplot(up_eWP,showCategory=10)
down_eWP@result$Description[1:10]
up_eWP@result$Description[1:10]
```

From the plot above we can see that most of the up-regulated genes are related to pathway connected to the cancer, like pancreatic cancer subtypes, cell cycle regulation. The down-regulated genes are related to pathway connected to complement system and coagulation cascades, 10q22q23 copy number variation and spinal cord injury. 
```{r, fig.width=10, fig.height=10, fig.align="center", echo = FALSE}
x2 <- pairwise_termsim(up_eWP) 
emapplot(x2)
```
From this plot related to up-regulated genes we can identify some networks, for example there are links between the matrix metalloproteinases and the chronic hyperglycemia impairment of neuron function, and between the retinoblastoma gene in cancer and the DNA replication and control of the cell cycle.

```{r, fig.width=10, fig.height=10, fig.align="center", echo = FALSE}
x2 <- pairwise_termsim(down_eWP) 
emapplot(x2)
```
From this plot related to up-regulated genes we can identify one network, there are links between Complement system in neuronal development and plasticity,its activation, and coagulation cascades. While these genes are not correlated to the 10q22q23 gene copy number variation.

**TASK 5:** Here we use pathview R package to visualize one particular pathway enriched using the up-regulated gene list

We decided to plot the pathway "Gastric cancer network 1" which we found enriched in the previous analysis. We used the code *hsa05226* which is the code number representing this KEGG pathway.

```{r}
library(pathview)

logFC <- up_DEGs$logFC
names(logFC) <- up_DEGs$entrezgene_id

data(demo.paths)
pv.out<-pathview(gene.data = logFC, 
         pathway.id = "hsa05226",
         species = "hsa", kegg.native = T)
head(pv.out$plot.data.gene)
```

The downstream analysis focuses only on down regulated genes (we made this choice for computational convenience since the down regulated ones are less).

**TASK 6:** In this step we identify the Transcription Factors which have enriched scores in the promoter of down regulated genes. The promoters are selected using the upstream sequence (window of 500nt) for the
selected genes.

```{r}
promoter_seq <- getSequence(id = down_DEGs$ensembl_gene_id, 
                            type="ensembl_gene_id",
                            seqType="gene_flank",
                            upstream=500, 
                            mart=ensembl)

sequences <- lapply(promoter_seq$gene_flank,function(x) DNAString(x))
length(sequences) 
```

We obtain 874 promoter sequences for the down regulated genes.

The motif enrichment has been calculated using the position weight matrix "PWMLogn.hg19.MotifDb.Hsap". To save computational time the result of the function *motifEnrichment()* has been saved in *enriched_transcription.RData*.

```{r}
data(PWMLogn.hg19.MotifDb.Hsap)
#enriched_TFs <- motifEnrichment(sequences,PWMLogn.hg19.MotifDb.Hsap,score = "affinity")
#save(enriched_TFs, file = "enriched_transcription.RData")
load("enriched_transcription.RData")
```

We then used the *sequenceReport()* function that create a ranked list of motifs

```{r}
report = groupReport(enriched_TFs)
```

```{r, fig.width=10, fig.height=10, fig.cap="Plot of the PWMs for the first top 5 ranked target", fig.align="center", echo = FALSE}
plot(report[1:5], fontsize=7, id.fontsize=5) #primi 5 TFs
```

**TASK 7:** Select one among the top enriched TFs, compute the empirical distributions of scores for all PWMs that you find in MotifDB for the selected TF and determine for all of them the distribution (log2) threshold cutoff at 99.75%.

```{r}
tfs = report$target[1] #select top target
tfs
```

The target we will consider is PGAM2 (Phosphoglycerate Mutase 2)

```{r}
tfs_motifs = subset(MotifDb, organism=='Hsapiens' & geneSymbol==tfs) #1 pos-fr-matrix
PWM = toPWM(as.list(tfs_motifs)) #position weight matrix for PGAM2
names(PWM) <-  sapply(names(PWM),function(x) strsplit(x,"-")[[1]][3])
```

Here we calculate the empirical distributions of scores for the only one PWM found and we use a threshold-cutoff at 99.75%

```{r}
ecdf = motifEcdf(PWM,organism = "hg19",quick=TRUE) 
threshold = log2(quantile(ecdf$PGAM2, 0.9975)) #compute the threshold
```

Here we use the *motifScores()* function to obtain the raw scores at each position in the sequences.

```{r}
scores = motifScores(sequences,PWM,raw.score=TRUE,cutoff=unlist(threshold))
```


Use the *plotMotifScores()* function to plot the log2 scores over the first 5 sequence. The motif hits are shown as rectangles with the base
being the length of the motif, and the height being the log2 score of the motif hit. The forward strand hits are shown on the top, and the reverse
strand hits are shown on the bottom.

```{r, fig.width=10, fig.height=10,fig.cap = "Plot of motif scores of the first 5 sequences for the motif PGAM2", fig.align="center", echo = FALSE}
plotMotifScores(scores[1:5], sel.motifs = "PGAM2", col = "red", cutoff = threshold)
```

**TASK 8:**
Starting from the list of the TFs we got before, we want to identify which of the down-regulated genes have a binding site in their upstream sequence for the TFs. We will only consider sites with a threshold above of the one calculated above. We will use Pattern Matching.


```{r}
#TASK 8: which down-reg genes have a region in their promoter with binding scores above the threshold for any of the selected PWMs

TFs <- report$target
#TFs <- TFs[!grepl("^UW.M", TFs)]
fractions_99_9<-c()
scoress <- data.frame(c(1:length(sequences)))
for (i in (1:length(TFs))){
  TFs_motif <- subset(MotifDb, organism == "Hsapiens" & geneSymbol == TFs[i])
  PWMs <- toPWM(as.list(TFs_motif))
  names(PWMs) <-  sapply(names(PWMs),function(x) strsplit(x,"-")[[1]][3])
  ecdfs <- motifEcdf(PWMs, organism = "hg19", quick = T)
  threshold <-lapply(ecdfs,function(x) log2(quantile(x,0.9975)))
  name <- TFs[i]
  score <- motifScores(sequences, PWMs, raw.scores = FALSE, cutoff = unlist(threshold))
  fractions_99_9 = c(fractions_99_9,length(which(apply(score,1,sum)>0))/874) #how many sequences for each TFs have a total score higher than zero
  sco <- as.data.frame(apply(score, 1, sum)) #I compute the total score for each sequence
  colnames(sco) <- name
  length(sequences)
  scoress<- cbind(scoress, sco)

}
save(scoress, file = "Scorestable.RData") #874 rows ( sequences) and target TFs columns
load("Scorestable.RData")
dim(scoress)
```
In the file *Scorestable.RData* we saved the score of each TFs for each upstream sequence of the genes. At this point we filter the TFs based on these criteria :
-if a TF has null scores for all the sequences we won't consider it, same for sequences which have 0 values for every TFs.
-We will keep only the sequences which have the greatest number of TFs binding region with a score higher than the median value of the score distribution
```{r}
#Since the first column contains the length of each sequence we will remove it
scoress <- as.matrix(scoress[, 2:length(scoress)])
head(scoress[, 1:10])
dim(scoress)  #524 TFs

# filter out TFs that don't have any score greater than 0 for any of the sequences
for (i in (1:length(fractions_99_9))){
     if (fractions_99_9[i] <= 0/874){
    col <- i
    scoress<-scoress[,-col]
     }
}
dim(scoress)
for (i in (1:length(sequences))) {
  if (sum(scoress[i])<0){
    scoress<-scoress[-i,]
  }
 }
dim(scoress)
```
After removing the zero columns we can see that no-TFs have a 0 score for every of the 874 sequences, same for sequences. Here we analyze the empirical distribution of each TFs, and we will use the median as a threshold in order to choose which down-regulated genes have a region in their promoter for a specific TFs
```{r}
# our threshold will be the median of each distribution

thresholds <- c()

for (i in(1:ncol(scoress))){
  thresholds<-c(thresholds,median(scoress[,i]))

}
#here I have a vector with the median of each distribution of each TFs scores
#I will consider that an upstream-sequence have a region binding a TFs if its score is over the median for its distribution
for (i in (1:ncol(scoress))){
  for (u in (1:nrow(scoress))){
    if (scoress[u,i]<thresholds[i]){
      scoress[u,i]<-0
    }
  }
}
countTFsperseq<-c()

for (i in (1:nrow(scoress))){
  countTFsperseq<-c(countTFsperseq,length(which(scoress[i,]!=0))) #
}
```
This vector contains, for each upstream sequence of the down-regulated genes, how many TF-binding regions are present.
Specifically, for the TFs we had choosen as a target before, we can see how many sequences does it bind :
```{r}
length(which(scoress[,"PGAM2"]!=0))

```
It binds 479 upstream-sequences

**TASK 9:** Use STRING database to find protein-protein interactions
among differentially expressed genes.

We select the most significant genes from task 8 and upload the list on
the online database STRING in order to find the PPI (the results are in
the file *string_interactions_short.tsv*)

```{r}
#top_enrich <- sequences$ensembl_gene_id[result]
top_enrich <- promoter_seq$ensembl_gene_id
down_names <- down_DEGs$external_gene_name[which(top_enrich%in%down_DEGs$ensembl_gene_id)]
#write.table(down_names,file="down_names.txt",row.names=F,col.names=T,sep="\t",quote=F)
links <- read.delim("string_interactions_short.tsv")
```

**TASK 10:** Import the network in R and using igraph package and
identify and plot the largest connected component.

First retrieve the annotations for building the nodes using biomaRt

```{r}
ensembl <- useMart(biomart="ensembl",dataset="hsapiens_gene_ensembl")

nodes <- 
  getBM(attributes=c("external_gene_name","ensembl_gene_id","description","gene_biotype","start_position","end_position","chromosome_name","strand"),
        filters=c("ensembl_gene_id"), 
        values=top_enrich,
        mart = ensembl)
```

Select the edges corresponding to the genes present in the nodes.
```{r}
nodes <-  unique(nodes[,c(1,3:6)]) %>% arrange(external_gene_name)
index <- links$node2%in%nodes$external_gene_name
links <- links[index,]
index <- links$X.node1%in%nodes$external_gene_name
links <- links[index,]
```

Create the network and simplify it removing multiple edges.

```{r}
net <- graph_from_data_frame(d=links,vertices=nodes,directed=FALSE)
net <- simplify(net)
```

Identify and plot the largest connected component

```{r, fig.width = 10, fig.height = 10,fig.cap="Network plot", fig.align = "center", echo = FALSE}
c <- components(net, mode =  c("weak","strong"))
# find max and position
max(c$csize) #strano perchÃ¨ sono tutti a 1 
net_work <- which.max(c$csize)
net.c <-  induced_subgraph(net,V(net)[which(c$membership == which.max(c$csize))])
col = rep("orange",length(V(net)))
plot(net.c, 
     edge.width=5,
     vertex.color=col,
     vertex.size=log10(as.numeric(V(net.c)$end_position-V(net.c)$start_position)),
     vertex.frame.color="darkgray",
     vertex.label.color="black", 
     vertex.label.cex=0.7,
     edge.curved=0.1) 
```

Taking into consideration only this extracted nodes a clustering have
been performed using the edges as a reference to build the clusters

Build clusters

```{r, fig.width = 10, fig.height = 10, fig.cap="Network plot highlighting the clusters",fig.align = "center", echo = FALSE}
ceb <- cluster_edge_betweenness(net.c)#find the communities
plot(ceb, net.c, vertex.size = 10)
#dendPlot(ceb, mode="hclust")

# retrieving the names from each gene list
#hclust_obj <- as.hclust(ceb)
#clusters <- cutree(hclust_obj, k = 11)
#cluster_names <- lapply(unique(clusters), function(cluster) {
 # names(clusters)[clusters == cluster]
#}) #nomi dei geni raggruppati in base ai cluster
#for (i in seq_along(cluster_names)) {
 # cat("Cluster", i, ": ", paste(cluster_names[[i]], collapse = ", "), "\n")
#}

```
